barman/postgres.py

```diff
@@ -1813,3 +1839,72 @@ def stop_exclusive_backup(self):
         return self._stop_backup(
             super(StandbyPostgreSQLConnection, self).stop_exclusive_backup
         )
+
+
+class PostgresKeepAlive:
+    """
+    Context manager to maintain a Postgres connection alive.
+
+    A child thread is spawned to execute heartbeat queries in the background
+    at a specified interval during its living context.
+
+    It does not open or close any connections on its own. Instead, it waits for the
+    specified connection to be opened in the main thread, making use of the same
+    current transaction.
+    """
+
+    THREAD_NAME = "barman_keepalive_thread"
+
+    def __init__(self, postgres, interval):
+        """
+        Constructor.
+
+        :param barman.postgres.PostgreSQLConnection postgres: The
+            Postgres connection to keep alive.
+        :param int interval: An interval in seconds at which a
+            heartbeat query will be sent to keep the connection alive.
+            A value <= 0 won't start the keepalive.
+        """
+        self.postgres = postgres
+        self.interval = interval
+        self._stop_thread = threading.Event()
+        self._thread = threading.Thread(
+            target=self._run_keep_alive,
+            name=self.THREAD_NAME,
+        )
+
+    def _sleep(self):
+        """
+        This is a custom sleep mechanism to avoid unecessary waits e.g. consider
+        the following scenario: 1) An interval of 300s is set 2) It executes a
+        hearbeat query and start sleeping for 300 seconds directly 3) After 100
+        seconds, the main tread finishes its job and is ready to exit the context
+        manager 4) Now the main thread have to wait for more 200 seconds until this
+        thread ends sleeping, even though all the work is already finished. To solve
+        this, instead of sleeping 300 seconds directly, we sleep 1 second 300 times
+        and each time we check if the thread should still be alive.
+        """
+        for _ in range(self.interval):
+            if self._stop_thread.is_set():
+                return
+            time.sleep(1)
+
```

Comment:
What about using `self._stop_thread.wait(self.interval)` instead?
That method does what you are trying to accomplish here, i.e. "sleep for N seconds unless something wakes it up" (`set` wakes it up, see https://docs.python.org/3/library/threading.html#threading.Event.set).

Reply:
Yes, I'd say to use the mechanism exposed by the `threading` module. The code will essentially do something similar, but in a cleaner way.

Reply:
We can remove the function and directly call `.wait(self.interval)` in `_run_keep_alive`.